Cython/Compiler/ModuleNode.py

```diff
@@ -262,7 +262,7 @@ def h_entries(entries, api=0, pxd=0):
             api_guard = self.api_name(Naming.api_guard_prefix, env)
             h_code_start.putln("#ifndef %s" % api_guard)
             h_code_start.putln("")
-            self.generate_extern_c_macro_definition(h_code_start)
+            self.generate_extern_c_macro_definition(h_code_start, env)
```

Comment:
I think (!) the header file isn't necessarily C++ specific even when we generate C++ code, so this should better not look at `env.is_cpp()` but use pure compile-time C/C++ features.

Reply:
My view was:
* if you're passing any C++ class (and a struct can be a hidden a C++ class) then the header is C++ specific. It's definitely very hard for Cython to tell if it's C++ specific.
* The user can always redefine the `__PYX_EXTERN_C` macro themselves, so we aren't forcing them into being C++ specific.
* With that said, I'm wondering what's wrong with declaring the functions as plain `extern` in both cases. There possibly is something I'm missing though